### PR TITLE
lb-workspace

### PR DESCRIPTION
**Pull request by Parth**
# Release blockers

+ [ ] horizontal scroll event
+ [x] egui app not initial syncing -- @ad-tra 
+ [x] Ugly -- @ad-tra 
+ [ ] Possible focus issue -- @tvanderstad / @ad-tra 
+ [x] Integration not sent to swift -- @Parth 
+ [x] Workspace current file not sent to swift -- @Parth 
+ [x] Sync status not sent to swift -- @Parth 
+ [ ] automatic renaming of files not working
+ [ ] settings hiding toolbar regression on egui

> its hard to decipher that the title of the page is the "title" bc it's the same font and styling as the rest of the text

# Merge blockers

+ [x] iOS integration -- @smailbarkouch 
+ [ ] Apple compatible pdf solution -- @Parth 

# Content Pane

The editor experiment has been a success, and this PR further extends the gains made by bringing more egui widgets to Apple.

+ We want to support many file types, and do not want to implement a custom editor for each file type on each platform. We want to implement it once and bring it to many places. Egui has already implemented support for the following file types:
  + Drawing
  + PDFs
  + Images
  + Plain Text
+ We also need to reduce the amount of duplicated effort we're performing. Quickly implementing these areas on various platforms served as a good proof of concept. But there are bugs and we can't afford the time to fix those bugs and unify the experience and create more complicated experiences for every platform.
  + tab strips
  + toolbars & shortcuts
  + auto save & auto sync

# Contract

Clients just tell content pane about file tree interactions (this UUID was clicked).

The following happens within content pane:
+ determining how to open a file and what to show
+ saving data
+ initial file names
+ determining when to kick off a sync
+ determining what content needs to be reloaded
+ tabs
+ toolbars

Content pane just needs to report back what file is currently selected (for nice file tree interactions). It can additionally, in the future when clients are ready, report the various tabs that are open and their dirtiness.

Content pane additionally probably needs to report back regarding whether it wants the software keyboard (and or styluses, etc).

# Considerations

We've had lots of conversations about the balance between native & cross platform UIs. The piece of this that gives me some pause is the tab strip. And primarily how it relates to #2152. However I think our current approach and it's future is on par with what we have and can achieve with egui. I do however really like the way macOS (not SwiftUI) native tabs feel, and would consider interacting with those in some future iteration of this work.

However, an important performance consideration, we don't want to initialize a content pane for every tab. We want mutually exclusive tabs (tabs on the same window) to share an underlying graphical instance. And only have multiple graphical instances for unique windows. I think this would require a medium-large sized technical lift no matter what approach we took. For this we may be in the territory of managing our own drag and drop events across two egui instances.

However, before we can even think about doing that, we need to more clearly understand what flow of editing the same note in two different windows would be. And so I think we can deliver 85% of the value of #2152 by deploying the following UX in the medium term:

You click on a note in a file tree, it opens it in a tab in the main window, whether that's a transient tab (#2104 / #2103) or not doesn't matter. You double click, and it will pop open a different window with just that note, you can tile it using your native window manager or whatever. If that note was already present in a tab you would be popping it out. Potentially there could be a pop out / pop in button or keyboard shortcut in either place.

The only workflow this solution is not compatible with is: viewing the same note side by side. But otherwise I think this is going to be our local optima for: team productivity, overall quality / reliability, and achieving the essence of that particular feature request.

# vague status

+ [x] pane
+ [x] pane-ffi
+ [x] eliminate ffi out of editor
+ [x] rework infra to reference "pane" instead of "editor"
+ [x] conform to contract inside rust and egui 
+ [x] conform to contract inside apple
+ [x] certainly a good amount of iOS work
+ [x] pdfium macos
+ [x] pdfium iOS
+ [x] possibly stylus events

### CODE REVIEW

**Comment by tvanderstad** [comment]
Can you clarify what’s in the MVP vs what’s not? Checklist here is in the right direction.

Can toolbars be done as a later step? Surely multiple windows can be done as a later step as well - at least how Windows is looking, multiple window support is a sizable lift.

File: clients/apple/macOS/FileListView.swift [hunk 1]
```diff
@@ -71,58 +73,58 @@ struct FileListView: View {
...
-                
-                Button(action: {
-                    currentSelection.isPendingSharesOpen = true
-                }) {
-                    pendingShareToolbarIcon(isPendingSharesEmpty: share.pendingShares.isEmpty)
-                        .imageScale(.large)
-                }
-            }
-        }
+//        .toolbar {
```
Comment (Parth) [inline clients/apple/macOS/FileListView.swift:89]: @smailbarkouch can you restore this

File: libs/content/workspace-ffi/SwiftWorkspace/Sources/Editor/macMTK.swift [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+           let url = createTempDir() {
+            let imageUrl = url.appendingPathComponent(String(UUID().uuidString.prefix(10).lowercased()), conformingTo: .png)
+            
+            do {
+                try data.write(to: imageUrl)
+            } catch {
+                return false
+            }
+            
+            //            if let lbImageURL = workspaceState!.importFile(imageUrl) {
```
Comment (Parth) [inline libs/content/workspace-ffi/SwiftWorkspace/Sources/Editor/macMTK.swift:181]: what's happening here @smailbarkouch?

File: libs/content/workspace-ffi/SwiftWorkspace/Sources/Editor/macMTK.swift [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+            //                paste_text(wsHandle, lbImageURL)
+            //                workspaceState?.pasted = true
+            //
+            //                return true
+            //            }
+        } else if isDropOperation {
+            if let data = pasteBoard.data(forType: .fileURL) {
+                if let url = URL(dataRepresentation: data, relativeTo: nil) {
+                    //                    if let markdownURL = workspaceState!.importFile(url) {
```
Comment (Parth) [inline libs/content/workspace-ffi/SwiftWorkspace/Sources/Editor/macMTK.swift:190]: also here @smailbarkouch

File: libs/content/workspace/src/workspace.rs [hunk 4]
```diff
@@ -270,44 +221,43 @@ impl super::AccountScreen {
...
                 egui::Stroke { color: ui.visuals().widgets.active.bg_fill, ..Default::default() };
             ui.visuals_mut().widgets.hovered.fg_stroke =
                 egui::Stroke { color: ui.visuals().widgets.active.bg_fill, ..Default::default() };
-            if Button::default().text("New folder").show(ui).clicked() {
-                self.update_tx
-                    .send(OpenModal::NewFolder(None).into())
-                    .unwrap();
-                ui.ctx().request_repaint();
-            }
+            // if Button::default().text("New folder").show(ui).clicked() {
```
Comment (Parth) [inline libs/content/workspace/src/workspace.rs:230]: parth restore this

File: libs/content/workspace/src/workspace.rs [hunk 5]
```diff
@@ -410,123 +358,321 @@ impl super::AccountScreen {
...
+            .name;
+
+        let fpath = self.core.get_path_by_id(id).unwrap(); // TODO
+
+        self.open_tab(id, &fname, &fpath, is_new_file);
+
+        let core = self.core.clone();
+        let ctx = self.ctx.clone();
+
+        // todo
```
Comment (Parth) [inline libs/content/workspace/src/workspace.rs:416]: restore this, but not for macOS

File: libs/lb/lb_external_interface/src/c_interface.rs [hunk 6]
```diff
@@ -731,10 +731,32 @@ pub unsafe extern "C" fn time_ago(time_stamp: i64) -> *const c_char {
 /// Be sure to call `release_pointer` on the result of this function to free the data.
 #[no_mangle]
 pub unsafe extern "C" fn get_core_ptr() -> *mut c_void {
-    let obj = static_state::get().expect("Could not get core");
+    let obj = static_state::get().expect("Could not get core").core;
     Box::into_raw(Box::new(obj)) as *mut c_void
 }
 
+struct CameraHandle(*mut c_void);
```
Comment (Parth) [inline libs/lb/lb_external_interface/src/c_interface.rs:738]: parth remove this

File: libs/content/workspace-ffi/SwiftWorkspace/Sources/Editor/macMTK.swift (hunk 2 above)
Comment (smailbarkouch) [reply libs/content/workspace-ffi/SwiftWorkspace/Sources/Editor/macMTK.swift:181]: Weird, I already fixed all of paste import. I think someone forced pushed it out of the history, probably when we were trying to fix merge conflicts.

File: clients/apple/Shared/Stateful Logic/StatusService.swift [hunk 7]
```diff
@@ -20,17 +20,18 @@ class StatusService: ObservableObject {
...
-            DispatchQueue.main.async {
-                switch localChanges {
-                case .success(let local):
-                    self.work = local.count
-                case .failure(let err):
-                    DI.errors.handleError(err)
-                }
-            }
-        }
+        // this whole file needs to be deleted
```
Comment (Parth) [inline clients/apple/Shared/Stateful Logic/StatusService.swift:23]: Delete this whole file eventually

File: clients/apple/Shared/Stateful Logic/SyncService.swift [hunk 8]
```diff
@@ -113,7 +92,7 @@ class SyncService: ObservableObject {
     }
     
     func sync() {
-        if syncing {
+        if true {
```
Comment (Parth) [inline clients/apple/Shared/Stateful Logic/SyncService.swift:95]: Delete this function

**Pull request merged**